From ebc0bec8f5c0f9ea4158f5e9f68c28a466b612c1 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Wed, 4 Sep 2019 21:13:45 +0200
Subject: [PATCH 1/1] efi_loader: cursor positioning

When backspacing in column 0 do no set the column index to ULONG_MAX.
Ensure that the row number is not set to ULONG_MAX even if the row count is
advertised as 0.
Ignore control characters other the 0x08, 0x0a, 0x0d when updating the
column.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 lib/efi_loader/efi_console.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/lib/efi_loader/efi_console.c b/lib/efi_loader/efi_console.c
index d4765afb98..4d8bc97998 100644
--- a/lib/efi_loader/efi_console.c
+++ b/lib/efi_loader/efi_console.c
@@ -156,13 +156,14 @@ static efi_status_t EFIAPI efi_cout_output_string(
 	 * Update the cursor position.
 	 *
 	 * The UEFI spec provides advance rules for U+0000, U+0008, U+000A,
-	 * and U000D. All other characters, including control characters
-	 * U+0007 (BEL) and U+0009 (TAB), have to increase the column by one.
+	 * and U000D. All other control characters are ignored. Any non-control
+	 * character increase the column by one.
 	 */
 	for (p = string; *p; ++p) {
 		switch (*p) {
 		case '\b':	/* U+0008, backspace */
-			con->cursor_column = max(0, con->cursor_column - 1);
+			if (con->cursor_column)
+				con->cursor_column--;
 			break;
 		case '\n':	/* U+000A, newline */
 			con->cursor_column = 0;
@@ -171,6 +172,8 @@ static efi_status_t EFIAPI efi_cout_output_string(
 		case '\r':	/* U+000D, carriage-return */
 			con->cursor_column = 0;
 			break;
+		case 0x00 ... 0x1f:
+			break;
 		case 0xd800 ... 0xdbff:
 			/*
 			 * Ignore high surrogates, we do not want to count a
@@ -185,6 +188,8 @@ static efi_status_t EFIAPI efi_cout_output_string(
 			con->cursor_column = 0;
 			con->cursor_row++;
 		}
+		if (con->cursor_row >= mode->rows && con->cursor_row)
+			con->cursor_row--;
 		con->cursor_row = min(con->cursor_row, (s32)mode->rows - 1);
 	}
 
-- 
2.23.0.rc1

